Teach IRParser to parse strides along with sizes in a tensor type.

ZolotukhinM opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #36952 [WIP][TensorExpr] kernel.cpp cleanup
* #36686 [TensorExpr] Remove obsolete code for handling dynamic shapes from kernel.cpp.
* **#36951 Teach IRParser to parse strides along with sizes in a tensor type.**

Differential Revision: [D21139940](https://our.internmc.facebook.com/intern/diff/D21139940)

Commit:
Teach IRParser to parse strides along with sizes in a tensor type.

[ghstack-poisoned]

Commit:
rebase on "Teach IRParser to parse strides along with sizes in a tensor type."


Differential Revision: [D21139940](https://our.internmc.facebook.com/intern/diff/D21139940)

[ghstack-poisoned]

Krovatkin commented on torch/csrc/jit/frontend/schema_type_parser.cpp:
```diff
@@ -165,18 +165,35 @@ TypePtr SchemaTypeParser::parseRefinedTensor() {
-          "Bad tensor dimension size. Strides not yet supported in parsing",
-          num);
       dims.push_back(dim);
+      if (L.cur().kind == ':') {
```
what happens if some of the dimensions are missing the stride information? 
can we enforce that we either have the stride information for all dimensions or we don't?

ZolotukhinM replied:
Initially I thought that it could be a user error, but thinking about it more I decided to detect and report it as you suggested. I've updated the PR accordingly!

Krovatkin reviewed (approved):
:shipit:

Commit:
Update on "Teach IRParser to parse strides along with sizes in a tensor type."


Differential Revision: [D21139940](https://our.internmc.facebook.com/intern/diff/D21139940)

[ghstack-poisoned]

suo commented:
If a graph has stride information, will its printed form (`dump()`) use this format?

ZolotukhinM commented:
> If a graph has stride information, will its printed form (`dump()`) use this format?

Currently the strides info will be ignored by the printer IIUC (and thus, roundtrip-check with stride-info will not be possible even with this PR).

suo commented:
Can we write the printing logic as well then? Seems like a pretty convenient thing to have (how are you going to get those graphs to parse, after all 😛). Getting toward round-trippability is good in general too.

Commit:
Fix test_jit/test_gpu_fusion on "Teach IRParser to parse strides along with sizes in a tensor type."


Differential Revision: [D21139940](https://our.internmc.facebook.com/intern/diff/D21139940)

[ghstack-poisoned]

Commit:
Teach IR printer to show strides too. on "Teach IRParser to parse strides along with sizes in a tensor type."


Differential Revision: [D21139940](https://our.internmc.facebook.com/intern/diff/D21139940)

[ghstack-poisoned]

ZolotukhinM commented:
> Can we write the printing logic as well then? Seems like a pretty convenient thing to have (how are you going to get those graphs to parse, after all 😛). Getting toward round-trippability is good in general too.

Sure, I've added this too. I think there is still more work to do here to get to an ideal state: e.g. we're not showing device and requires_grad info, but I'll leave this part for now (it might be an easy onboarding task btw).

Commit:
Update on "Teach IRParser to parse strides along with sizes in a tensor type."


Differential Revision: [D21139940](https://our.internmc.facebook.com/intern/diff/D21139940)

[ghstack-poisoned]
